packages/playground/data-liberation/src/import/WP_Topological_Sorter.php

```diff
@@ -0,0 +1,198 @@
+			'visited'     => false,
+		);
+	}
+
+	public function map_post( $byte_offset, $data ) {
+		if ( empty( $data ) ) {
+			return false;
+		}
+
+		// No parent, no need to sort.
+		if ( ! isset( $data['post_type'] ) ) {
+			return false;
+		}
+
+		if ( 'post' === $data['post_type'] || 'page' === $data['post_type'] ) {
+			if ( ! $data['post_id'] ) {
+				$this->last_post_id = $this->orphan_post_counter;
+				--$this->orphan_post_counter;
+			}
+
+			// This is an array saved as: [ parent, byte_offset ], to save
+			// space and not using an associative one.
+			$this->posts[ $data['post_id'] ] = array(
+				$data['post_parent'],
+				$byte_offset,
+			);
+		}
+
+		return true;
+	}
+
+	/**
+	 * Get the byte offset of an element, and remove it from the list.
+	 */
+	public function get_byte_offset( $id ) {
+		if ( ! $this->sorted ) {
+			return false;
+		}
+
+		if ( isset( $this->posts[ $id ] ) ) {
+			$ret = $this->posts[ $id ];
+
+			// Remove the element from the array.
+			unset( $this->posts[ $id ] );
+
+			if ( 0 === count( $this->posts ) ) {
+				// All posts have been processed.
+				$this->reset();
+			}
+
+			return $ret;
+		}
+
+		return false;
+	}
+
+	public function is_sorted() {
+		return $this->sorted;
+	}
+
+	/**
+	 * Sort posts topologically.
+	 *
+	 * Children posts should not be processed before their parent has been processed.
+	 * This method sorts the posts in the order they should be processed.
+	 *
+	 * Sorted posts will be stored as attachments and posts/pages separately.
+	 */
+	public function sort_topologically( $free_space = true ) {
+		foreach ( $this->categories as $slug => $category ) {
+			$this->topological_category_sort( $slug, $category );
+		}
+
+		$this->sort_elements( $this->posts );
+
+		// Free some space.
+		if ( $free_space ) {
+			/**
+			 * @TODO: all the elements that have not been moved can be flushed away.
+			 */
+			foreach ( $this->posts as $id => $element ) {
+				// Save only the byte offset.
+				$this->posts[ $id ] = $element[1];
+			}
+		}
+
+		$this->sorted = true;
+	}
+
+	/**
+	 * Recursive sort elements. Posts with parents will be moved to the correct position.
+	 *
+	 * @return true
+	 */
+	private function sort_elements( &$elements ) {
+		$sort_callback = function ( $a, $b ) use ( &$elements ) {
+			$parent_a = $elements[ $a ][0];
+			$parent_b = $elements[ $b ][0];
+
+			if ( ! $parent_a && ! $parent_b ) {
+				// No parents.
+				return 0;
+			} elseif ( $a === $parent_b ) {
+				// A is the parent of B.
+				return -1;
+			} elseif ( $b === $parent_a ) {
+				// B is the parent of A.
+				return 1;
+			}
+
+			return 0;
+		};
+
+		/**
+		 * @TODO: PHP uses quicksort: https://github.com/php/php-src/blob/master/Zend/zend_sort.c
+		 * WordPress export posts by ID and so are likely to be already in order.
+		 * Quicksort performs badly on already sorted arrays, O(n^2) is the worst case.
+		 * Let's consider using a different sorting algorithm.
+		 */
+		uksort( $elements, $sort_callback );
```

Comment:
This requires fitting all `$elements` into memory. This may be fine for v1, since we're talking about relatively small data structures, but we'll hit the limits of this approach sooner than later. Is it possible to perform topological sorting with at a reasonable speed without holding everything in memory?

Reply:
We can save the data in a custom DB table (Jetpack have a custom `wp_jetpack_sync_queue` table) instead of a simple array in memory. I didn't disturb the database in this first phase, but we can. Reproducing this in a table and using the custom sort here is straightforward. At the end of collecting the byte offsets, it's a matter of `ALTER TABLE wxr_sorting ORDER BY custom_sorting_like_the_one_above`, and we are done, even with our new reentrancy model.

When it finds a row in this table, the streamer can jump to the "correct" position (the one it should load before the current one) and proceed to the next one.

Reply:
> It may seem excessive for v1, but thinking about it more it seems like something that could have a ripple through the entire system if we don't account for streaming early on.

Agree. It's reasonable, let's proceed this way. 👍